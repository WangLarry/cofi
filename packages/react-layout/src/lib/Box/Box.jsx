/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import styled from 'styled-components';

const Element = styled.div`${props => props.styled}`;

const Box = ({ style, boxes = [], component: Component, props = {} }) => {
  const elementProps = typeof style === 'string' ? { styled: style } : { style };

  return Component 
    ? (<Component {...props} />) 
    : (<Element {...elementProps}>{boxes.map((box, index) => (<Box key={index} {...box} />))}</Element>);
};

export default Box;
