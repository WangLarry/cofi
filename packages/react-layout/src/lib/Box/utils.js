/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

export function iterateBoxes(boxes = [], func) {
  boxes.forEach((box) => {
    iterateBoxes(box.boxes, func);
    func(box);
  });
}
