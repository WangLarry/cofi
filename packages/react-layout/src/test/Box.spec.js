/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { shallow } from 'enzyme';
import Box from '../lib/Box/index';

describe('Box', () => {
  const Field = () => {};
  const rowStyle = { display: 'flex', flexDirection: 'row' };
  const columStyle = { display: 'flex', flexDirection: 'column' };
  it('Should render component - ok', () => {
    const props = {
      component: () => {},
      props: { id: 'firstName' },
    };
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });

  it('Should render box with style, ok', () => {
    const props = {
      style: rowStyle,
      boxes: [
        { component: Field, props: { id: 'firstName' } },
        { component: Field, props: { id: 'lastName' } },
      ],
    };
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });

  it('Should render nested boxes, ok', () => {
    const props = {
      style: rowStyle,
      boxes: [{
        style: columStyle,
        boxes: [
          { component: Field, props: { id: 'firstName' } },
          { component: Field, props: { id: 'lastName' } },
        ],
      }, {
        style: columStyle,
        boxes: [
          { component: Field, props: { id: 'address' } },
        ],
      }],
    };
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });

  function getComponent(props) {
    return (<Box {...props} />);
  }
});
