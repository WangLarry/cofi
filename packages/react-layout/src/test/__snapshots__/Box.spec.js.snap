// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box Should render box with style, ok 1`] = `
<styled.div
  style={
    Object {
      "display": "flex",
      "flexDirection": "row",
    }
  }
>
  <Box
    component={[Function]}
    key="0"
    props={
      Object {
        "id": "firstName",
      }
    }
  />
  <Box
    component={[Function]}
    key="1"
    props={
      Object {
        "id": "lastName",
      }
    }
  />
</styled.div>
`;

exports[`Box Should render component - ok 1`] = `
<component
  id="firstName"
/>
`;

exports[`Box Should render nested boxes, ok 1`] = `
<styled.div
  style={
    Object {
      "display": "flex",
      "flexDirection": "row",
    }
  }
>
  <Box
    boxes={
      Array [
        Object {
          "component": [Function],
          "props": Object {
            "id": "firstName",
          },
        },
        Object {
          "component": [Function],
          "props": Object {
            "id": "lastName",
          },
        },
      ]
    }
    key="0"
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  />
  <Box
    boxes={
      Array [
        Object {
          "component": [Function],
          "props": Object {
            "id": "address",
          },
        },
      ]
    }
    key="1"
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  />
</styled.div>
`;
